### PR TITLE
New network system

### PR DESCRIPTION
**Pull request by HookWoods**
Hello, I'm porting a new patch to yatopia about the network.
The network is a part of spigot that has never been too much modified and can be optimized with some kernel parameters (I may add TCP_FASTOPEN in the future or other params like this).

Actually this patch only enhanced the code readability of the ServerConnection and add a new beta feature:
IOUring, it's a new async interface for IO introduce in Kernel 5.4 in Linux (So only work on linux with kernel >= 5.4)
Netty is actually working on a project to port IOUring on it (https://github.com/netty/netty-incubator-transport-io_uring).
This is actually in beta so that's why there is a config to enable it or disable it, but can offer some performance on the network.

There is also a known issue with Zlib in Bungeecord that makes IOUring unusable cause it can decompress packet from it 
(https://github.com/netty/netty-incubator-transport-io_uring/issues/40).

### CODE REVIEW

**Review by MrIvanPlays — CHANGES_REQUESTED** [review]
no readability changes. minimal diff

File: patches/server/0067-New-Network-System.patch [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+             LazyInitVar lazyinitvar;
+ 
+             if (Epoll.isAvailable() && this.e.l()) {
+@@ -72,14 +86,17 @@ public class ServerConnection {
+                 oclass = NioServerSocketChannel.class;
+                 lazyinitvar = ServerConnection.a;
+                 ServerConnection.LOGGER.info("Using default channel type");
+-            }
++            }*/
++            // Yatopia end
```
Comment (MrIvanPlays) [inline patches/server/0067-New-Network-System.patch]: commenting out big chunks of code should be like this: 
```java
public String something() {
  String astring = "my name is anton";
  /* // Yatopia start - better method 
  return astring.substring(11, astring.length() - 1);
  */
  // some new code
  // Yatopia end
}
```

**Comment by MrIvanPlays** [comment]
overall try to minimise the diff as much as possible

**Comment by duplexsystem** [comment]
I'd like to implement velocity's VelocityNettyThreadFactory and some minor formatting improvements before merge

**Comment by Titaniumtown** [comment]
Hmm very interesting PR.

File: patches/server/0067-New-Network-System.patch [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+         </dependency>
+         <!-- Tuinity end - fix compile issue (cannot see new api) by moving netty include BEFORE server jar -->
++        <!-- Yatopia Start - Add IOUring beta support -->
++        <dependency>
++            <groupId>io.netty.incubator</groupId>
++            <artifactId>netty-incubator-transport-native-io_uring</artifactId>
++            <version>0.0.3.Final</version>
++            <classifier>linux-x86_64</classifier>
++        </dependency>
++        <!-- YAtopia end -->
```
Comment (Titaniumtown) [inline patches/server/0067-New-Network-System.patch]: Typo

File: patches/server/0067-New-Network-System.patch [hunk 3]
```diff
@@ -0,0 +1,286 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Hugo Planque <hookwood@pop-os.localdomain>
```
Comment (Titaniumtown) [inline patches/server/0067-New-Network-System.patch]: Your email here is a local address, you may want to fix that xD

**Comment by MrIvanPlays** [comment]
chunks of code are still not commented out properly

**Review by MrIvanPlays — DISMISSED** [review]
I like these changes. 
I am aware of channel factories for like ages but I'm just too afraid to touch anything with this code. Looks like you've done it.

@YatopiaMC/review-team I need 1 more person to review it.

**Review by MrIvanPlays — DISMISSED** [review]

**Review by duplexsystem — CHANGES_REQUESTED** [review]
1. I cannot connect at all with IOUring enabled the first or current build
2. I think we should port Velocity's VelocityNettyThreadFactory because it uses a netty thread pool which will probably work better with netty
3. we need to drop the log level of trying to using different network types to info

**Comment by MrIvanPlays** [comment]
> I think we should port Velocity's VelocityNettyThreadFactory because it uses a netty thread pool which will probably work better with netty

um this pr also uses netty pool. the thread factory quite literally doesn't matter. velocity's is just changing the name. same for ours.

>  we need to drop the log level of trying to using different network types to info

ill remove it

**Comment by duplexsystem** [comment]
> um this pr also uses netty pool. the thread factory quite literally doesn't matter. velocity's is just changing the name. same for ours.

No it uses com.google.common.util.concurrent.ThreadFactoryBuilder. VelocityNettyThreadFactory uses io.netty.util.concurrent.FastThreadLocalThread and java.util.concurrent.ThreadFactory and I frankly trust astei's judgement here more than yours.

**Comment by MrIvanPlays** [comment]
things left:
fix io uring
update patches.md

**Review by duplexsystem — DISMISSED** [review]

**Review by MrIvanPlays — CHANGES_REQUESTED** [review]
blocking cuz i want to do some changes before merge

**Review by MrIvanPlays — DISMISSED** [review]

**Review by duplexsystem — DISMISSED** [review]

File: patches/server/0067-Port-krypton.patch [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+ 
+diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
+index 08227ab446d6332af76491a063653f7f13f43560..c02d8f1f29662e6ce02c0a24209bf18d56f7b3a4 100644
+--- a/src/main/java/net/minecraft/server/NetworkManager.java
++++ b/src/main/java/net/minecraft/server/NetworkManager.java
+@@ -20,7 +20,18 @@ import io.netty.util.concurrent.GenericFutureListener;
+ import java.net.SocketAddress;
+ import java.util.Queue;
+ import javax.annotation.Nullable;
+-import javax.crypto.Cipher;
```
Comment (MrIvanPlays) [inline patches/server/0067-Port-krypton.patch]: readd that import - smaller diff

File: patches/server/0067-Port-krypton.patch [hunk 5]
```diff
@@ -0,0 +1,552 @@
...
++// Yatopia start
++import javax.crypto.SecretKey;
++
```
Comment (MrIvanPlays) [inline patches/server/0067-Port-krypton.patch]: empty line :eyes:

File: patches/server/0066-New-Network-System.patch [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
++        <dependency>
++            <groupId>io.netty.incubator</groupId>
++            <artifactId>netty-incubator-transport-native-io_uring</artifactId>
++            <version>0.0.3.Final</version>
++            <classifier>linux-x86_64</classifier>
++        </dependency>
++        <!-- Yatopia end -->
+         <dependency>
+             <groupId>io.papermc</groupId>
+             <artifactId>minecraft-server</artifactId>
```
Comment (Titaniumtown) [inline patches/server/0066-New-Network-System.patch:30]: Shouldn't these pom changes be in `0001-Modify-POM.patch`?

File: patches/server/0067-Port-krypton.patch [hunk 7]
```diff
@@ -0,0 +1,552 @@
...
+diff --git a/src/main/java/org/bukkit/craftbukkit/Main.java b/src/main/java/org/bukkit/craftbukkit/Main.java
+index 756be0886856aabc31e436f82948d3d069f66fef..ad0494b7eadfa8280af2e784a6507184a4fbbd9a 100644
+--- a/src/main/java/org/bukkit/craftbukkit/Main.java
++++ b/src/main/java/org/bukkit/craftbukkit/Main.java
+@@ -10,6 +10,8 @@ import java.util.List;
+ import java.util.concurrent.TimeUnit;
+ import java.util.logging.Level;
+ import java.util.logging.Logger;
++
++import io.netty.util.ResourceLeakDetector;
```
Comment (MrIvanPlays) [inline patches/server/0067-Port-krypton.patch]: // Yatopia comment and remove that empty line :eyes:

**Comment by Titaniumtown** [comment]
Crap, it seems I had old review pending, lmao. Seems they all still apply though xD

**Review by MrIvanPlays — CHANGES_REQUESTED** [review]

File: patches/server/0066-New-Network-System.patch (hunk 6 above)
Comment (MrIvanPlays) [reply patches/server/0066-New-Network-System.patch:30]: no

Comment (Titaniumtown) [reply patches/server/0066-New-Network-System.patch:30]: Why not? At least bumping the netty version should be.

Comment (MrIvanPlays) [reply patches/server/0066-New-Network-System.patch:30]: > Why not?

patch related stuff should be in the patch itself.

> At least bumping the netty version should be.

it is not a necessarity.

File: patches/server/0067-New-Network-System.patch (hunk 3 above)
Comment (HookWoods) [reply patches/server/0067-New-Network-System.patch]: I will do it but I don't care here, before I was working on my windows computer and everything was set up like my gpg key, ...

Comment (MrIvanPlays) [reply patches/server/0067-New-Network-System.patch]: only thing left to resolve and ill approve.

**Review by MrIvanPlays — DISMISSED** [review]

**Review by Titaniumtown — DISMISSED** [review]

**Comment by krolik-exe** [comment]
how do I check if IOUring is working, because i enabled it in the config and when i boot up it shows using EPOLL network type and showing that it's using velocity compression and chipher, and doesn't show any errors

**Comment by HookWoods** [comment]
Actually you need to set network compression to -1 and bungee to false to make it works, after you can enable the debug that will tell you why io uring is not working after you did all that thing before

**Comment by krolik-exe** [comment]
ok i will try this

**Comment by krolik-exe** [comment]
Thanks, it works now

**Pull request merged**